GSW/testUtility/testNS2.pyw

```diff
@@ -26,87 +26,148 @@ commandField = Entry(commandFrame, width=30, borderwidth=3, font=('Consolas',10)
+        overwrite = tkinter.messagebox.askyesno(title='Warning', message= warning)
+        if (not overwrite):
+            return
+    # write to file
+    file = open(filePath,'w', newline = '')
+    file.write(monitor.get('1.0','end-1c'))
+    updateLog( 'Output saved to ' + fileName)
+
+# clear output #
+def clearOutput(): # clears the output monitor
+    confirm = tkinter.messagebox.askyesno(title='Warning', message= 'Are you sure you want to clear the output monitor???')
+    if confirm:
+        monitor.configure(state ='normal')
+        monitor.delete('1.0', 'end')
+        monitor.configure(state='disabled')
+        updateLog('Output monitor cleared.')
 
+# ==== buttons ====
+button_sendCommand = Button(commandFrame, text='Send', command=sendCommand)
+root.bind('<Return>', sendCommand) # bind the enter key to send command
+button_scanPorts = Button(serialFrame, text='Scan For Serial Ports', command=scanPorts)
+button_updatePort = Button(serialFrame, text='Open Serial Port', bg='#c9ffba', command=openSerialPort)
+button_saveFile = Button(fileFrame, text='Save to File', command=saveFile)
+button_clearOutput = Button(monitorFrame, text='Clear', command=clearOutput)
 
 # ==== draw everything ====
 # draw command frame
 commandFrame.grid(row=0, column=0, padx=5, pady=5, sticky=W+E)
 commandField.grid(row=0, column=0, padx=5, pady=3, sticky=W+E)
 button_sendCommand.grid(row=0, column=1, sticky=W+E)
 
-# log frame
+# draw log frame
 logFrame.grid(row=1, column=0, padx=5, pady=5, sticky=N+S+E+W)
 log.pack(expand=True, fill=BOTH)
 
 # draw monitor frame
-monitorFrame.grid(row=0, column=1, padx=5, pady=5, sticky=N+W+E+S, rowspan=3)
+monitorFrame.grid(row=0, column=1, padx=5, pady=5, sticky=N+W+E+S, rowspan=4)
 monitor.pack(expand=True, fill=BOTH)
+check_autoScroll.pack(side='left')
+button_clearOutput.pack(side='right')
 
 # draw serial frame
 serialFrame.grid(row=2, column=0, padx=5, pady=5, sticky=W+E)
 portField.grid(row=0, column=0, sticky=W+E)
 button_updatePort.grid(row=0, column=1, sticky=W+E)
 button_scanPorts.grid(row=1, column=0, sticky=W+E, columnspan=2)
 
+# draw file frame
+fileFrame.grid(row=3, column=0, padx=5, pady=5, sticky=W+E)
+fileNameField.grid(row=0, column=0, sticky=W+E)
+button_saveFile.grid(row=0, column=1, sticky=W+E)
 
 # ==== start application ====
-teensy = serial.Serial(port='COM3', baudrate=9600, timeout=10/1000)
+# configure serial connection
+portOpen = False
+teensy = serial.Serial()
+teensy.baudrate = 19200
+teensy.timeout = 0.05
+
+scanPorts() # scan for serial ports
+updateLog('\nTeensy usually appears as \"USB Serial Device\"')
+updateLog('Enter the Teensy\'s COM port and click \"Open Serial Port\" to connect.')
 commandField.focus() # place cursor in command field
 
-serialThread = threading.Thread(target=readSerial)
-
 while True:
-    readSerial()
-    root.update()   
+    root.update()
+    if portOpen:
+        readSerial()
```

Comment:
I would also really like to see an autosave feature. Thermal testing is very stressful, I don't want them to lose data. Can you add an autosave() function that just prints the serial monitor to an autosave file, every ~5 min. Just something simple as a backup. Here's some psuedocode:

now = datetime.now()

if datetime.now() > (now + five_min)
   autosave_file = open(/autosave.txt/, 'w', newline = '')
   autosave_file.write(monitor.get('1.0','end-1c'))
   now = datetime.now()

Reply:
Even better, I made it so each line is written to a backup file in real time